pkg/controllers/tokentimeoutsync/tokentimeoutsync.go

```diff
@@ -0,0 +1,179 @@
+			globalChanged = c.globalTimeout != -1 && ((c.globalTimeout > 0 && newTimeout < c.globalTimeout) || (c.globalTimeout == 0 && newTimeout > 0))
+			c.globalTimeout = newTimeout
+		}
+	} else {
+		globalChanged = c.globalTimeout != 0
+		c.globalTimeout = 0
+	}
+
+	tokenList, err := c.accessTokens.List(ctx, metav1.ListOptions{LabelSelector: labels.Everything().String()})
+	if err != nil {
+		return err
+	}
+
+	for _, t := range tokenList.Items {
+		if err := c.updateToken(ctx, &t, globalChanged); err != nil {
+			// can't log token until we are sure only sha256: tokens exist
+			klog.Errorf("failed to upate token")
+		}
+	}
+
+	oauthClients, err := c.oauthClientLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, client := range oauthClients {
+		var timeout *time.Duration
+		if client.AccessTokenInactivityTimeoutSeconds == nil {
+			timeout = nil
+		} else {
+			t := time.Duration(*client.AccessTokenInactivityTimeoutSeconds) * time.Second
+			timeout = &t
+		}
+		c.clientTimeouts[client.Name] = timeout
+	}
+
+	return nil
+}
+
+func (c *tokenTimeoutSyncController) updateToken(ctx context.Context, token *oauthv1.OAuthAccessToken, globalChanged bool) error {
+	tokenClientName := token.ClientName
+	if len(tokenClientName) == 0 {
+		// client name is a required attribute, this should never happen
+		klog.Errorf("found a token without a client name")
+		return nil
+	}
+
+	clientTimeout, err := c.getOAuthClientTimeout(tokenClientName)
+	if err != nil {
+		// failed to get the client, don't update anything
+		return err
+	}
+	tokenCopy := token.DeepCopy()
+
+	storedTimeout, _ := c.clientTimeouts[tokenClientName]
+	if clientTimeout == nil && (globalChanged || storedTimeout != nil) {
+		// either client timeout changed and is now nil or global timeout changed
+		if applyTimeout(tokenCopy, c.globalTimeout) {
+			_, err = c.accessTokens.Update(ctx, tokenCopy, metav1.UpdateOptions{})
+		}
+		return err
+	}
+
+	if clientTimeout != nil && (storedTimeout == nil || time.Duration(*clientTimeout)*time.Second < *storedTimeout) {
+		// either the stored timeout was nil or was higher than the current one
+		if applyTimeout(tokenCopy, time.Duration(*clientTimeout)*time.Second) {
+			_, err = c.accessTokens.Update(ctx, tokenCopy, metav1.UpdateOptions{})
+		}
+		return err
+	}
+
+	// no changes
+	return nil
+}
+
+func applyTimeout(tokenCopy *oauthv1.OAuthAccessToken, timeout time.Duration) bool {
+	var changed bool
+	if timeout == 0 {
+		changed = tokenCopy.InactivityTimeoutSeconds != 0
+		tokenCopy.InactivityTimeoutSeconds = 0
+		return changed
+	}
+
+	// count a new timeout based on the current time
+	newTimeout := int32((time.Now().Sub(tokenCopy.CreationTimestamp.Time) + timeout) / time.Second)
+	if oldTimeout := tokenCopy.InactivityTimeoutSeconds; oldTimeout == 0 || newTimeout < tokenCopy.InactivityTimeoutSeconds {
+		tokenCopy.InactivityTimeoutSeconds = newTimeout
```

Comment:
this is wrong. We cannot reset all of them. We have to add the difference between new and old (global/client) value.